census_view.cpp

```diff
@@ -1897,6 +3148,203 @@ void CensusDVCView::UponPasteCensus(wxCommandEvent&)
+                int z = value_cast<int>(values[j]);
+                if(jdn_min <= z && z <= jdn_max)
+                    {
+                    ; // Do nothing: JDN is the default expectation.
+                    }
+                else if(ymd_min <= z && z <= ymd_max)
+                    {
+                    z = YmdToJdn(ymd_t(z)).value();
+                    values[j] = value_cast<std::string>(z);
+                    }
+                else
+                    {
+                    alarum()
+                        << "Invalid date " << values[j]
+                        << " for '" << headers[j] << "'"
+                        << " on line " << current_line << "."
+                        << LMI_FLUSH
+                        ;
+                    }
+                }
+            current_cell[headers[j]] = values[j];
+            }
+        current_cell.Reconcile();
+        current_cell.RealizeAllSequenceInput();
+        cells.push_back(current_cell);
+
+        status() << "Added cell number " << cells.size() << '.' << std::flush;
+        }
+
+    if(0 == current_line)
+        {
+        warning() << "No cells to paste." << LMI_FLUSH;
+        return;
+        }
+
+    if(!document().IsModified() && !document().GetDocumentSaved())
+        {
+        case_parms ().clear();
+        case_parms ().push_back(archetype);
+        class_parms().clear();
+        class_parms().push_back(archetype);
+        cell_parms ().swap(cells);
+        }
+    else if(configurable_settings::instance().census_paste_palimpsestically())
+        {
+        cell_parms().swap(cells);
+        // Cells that were copied from lmi have DOB and not IssueAge,
+        // so pasting them back in sets UseDOB. Force UseDOB for case
+        // and class defaults to prevent showing a UseDOB column with
+        // each cell set to "Yes".
+        for(auto& j : case_parms ()) {j["UseDOB"] = "Yes";}
+        for(auto& j : class_parms()) {j["UseDOB"] = "Yes";}
+        }
+    else
+        {
+        cell_parms().reserve(cell_parms().size() + cells.size());
+        std::back_insert_iterator<std::vector<Input>> iip(cell_parms());
+        std::copy(cells.begin(), cells.end(), iip);
+        }
+
+    auto const old_rows = grid_window_->GetNumberRows();
+    auto const old_cols = grid_window_->GetNumberCols();
+    auto const new_rows = grid_table_->GetRowsCount();
+    auto const new_cols = grid_table_->GetColsCount();
+
+    wxGridUpdateLocker grid_update_locker(grid_window_);
+    grid_window_->ClearSelection();
+
+    if(grid_window_->IsCellEditControlEnabled())
+        {
+        grid_window_->DisableCellEditControl();
+        }
+
+    if(old_rows != new_rows || old_cols != new_cols)
+        {
+        grid_window_->DeleteRows(0, old_rows);
+        grid_window_->DeleteCols(0, old_cols);
+        grid_window_->AppendRows(new_rows);
+        grid_window_->AppendCols(new_cols);
+        grid_table_->make_cell_number_column_read_only();
```

Comment:
As we always have the column 0, couldn't we call `DeleteCols(1, old_cols - 1)` and `AppendCols(new_cols - 1)` instead and avoid calling this here?